README.md
# Komrad Filter Editor

Это расширение предназначено для редактирования фильтров в SIEM Komrad. Пока функционал весьма ограничен и довольно примитивен.

This extension is designed for editing filters in SIEM Komrad. While the functionality is very limited and rather primitive.

## Возможности (Features)

- Получение списка фильтров (Getting a list of filters)
- Открытие фильтра в открытом текстовом окне (Opening the filter in an open text window)
- Загрузка фильтра обратно в SIEM (Uploading the filter back into SIEM)

Кратко опишу работу с расширением. 

Сначала необходимо в настройках указать адрес SIEM и учетные данные. Если используете логин и пароль, оставляйте поле токена пустым. Если используете токен, учетные данные могут быть любыми. 

First, you need to specify the SIEM address and credentials in the settings. If you use a username and password, leave the token field empty. If you use a token, the credentials can be anything.

![](./images/settings.png)

После найдите панели слева иконку расширения. 

After that, find the extension icon on the left panel.

![](./images/extension_activity_bar.png)

Если все данные указаные верно, фильтры будут загружены и видны на панели. Если навести курсор на фильтр, будут достумпны две опции: редактирование и загрузка. Первая - вставляет код фильтра в открытое текстовое окно. В случае если оно не открыто, можно использовать сочетания *Ctrl+N*. Вторая -- загружет все содержимое текстового окна как код фильтра.

If all the data is specified correctly, the filters will be loaded and visible on the panel. If you hover the cursor over the filter, two options will be available: edit and upload. The first one inserts the filter code into an open text window. If it is not open, you can use the combinations *Ctrl+N*. The second one loads the entire contents of the text window as a filter code.

![](./images/filter_explorer.png)

## Зависимости (Requirements)

Не знаю что тут писать, ввиду отсутствия опыта. Помимо необходимых вещей, которые загрузил VS Code Extension Generator, используется `node-fetch`. 

I don't know what to write here, due to lack of experience. In addition to the necessary things that VS Code Extension Generator loaded, `node-fetch` is used.

## Настройки расширения (Extension Settings)

Это расширение предоставляет следующие настройки:

- `komrad-filter-editor.targetUrl`: адрес вашего Комрада;
- `komrad-filter-editor.login`: Логин учетной записи;
- `komrad-filter-editor.password`: Пароль учетной записи;
- `komrad-filter-editor.token`: Вместо учетных данных можно указать сессионный токен.

This extension contributes the following settings:

- `komrad-filter-editor.targetUrl`: Your Komrad instance address;
- `komrad-filter-editor.login`: Login for your account;
- `komrad-filter-editor.password`: Password for your account;
- `komrad-filter-editor.token`: Set session token instead of login and password.

## Известные проблемы (Known Issues)

Эта ранняя версия, поэтому возможны различные непредсказуемые проблемы.

This is an early version, so various unpredictable problems are possible.

- На данный момент игнорируется проверка сертификата. Планируется добавить функцию позже. (Certificate verification is ignored at the moment. It is planned to add this feature later.)


## Примечания к выпуску (Release Notes)

### 0.0.1

Это начальный выпуск. Представлены некоторые основные функции.

This is the initial release. Some basic functions are presented.


CHANGELOG.md
# Журнал изменений (Change Log)

Тут пока пусто

It's still empty here

src/extension.ts
import * as vscode from 'vscode';
import fetch from 'node-fetch';
import { escape } from 'querystring';
import { buffer, json } from 'stream/consumers';

export async function activate(context: vscode.ExtensionContext) {

	console.log('Congratulations, your extension "komrad-filter-editor" is now active!');

	const config = vscode.workspace.getConfiguration('komrad-filter-editor');

	let cookie = config.get('token');
	if (!cookie || cookie === undefined || cookie === null) {
		cookie = await komradLogin();
	}

	if (cookie !== null) {
		const filterProvider = new FilterProvider(await getFiltersList(String(cookie)));

		let filterTree = vscode.window.registerTreeDataProvider('filters-list', filterProvider);

		let filtersList = vscode.commands.registerCommand('filters-list.refresh', () => {
			filterProvider.refresh(String(cookie));
		});

		let filterEdit = vscode.commands.registerCommand('filters-list.edit', (node: Filter) => {
			let editor = vscode.window.activeTextEditor;
			let sourceCode = Buffer.from(node.source, 'base64').toString('utf-8');
			if (editor) {
				let range = new vscode.Range(0, 0, editor.document.lineCount, 0);
				editor.edit(editBuilder => {
					editBuilder.replace(range, sourceCode);
				});
				vscode.window.showInformationMessage(`Фльтр ${node.label} успшно получен`);
			}
			else {
				vscode.window.showErrorMessage('Сначала откроте новое пустое окно (Ctrl+N)');
			}
		});

		let filterUpload = vscode.commands.registerCommand('filters-list.upload', (node: Filter) => {
			updateFilter(String(cookie), node);
		});

		context.subscriptions.push(filterTree, filtersList, filterEdit, filterUpload);
	}
}

export async function deactivate() { }

async function komradLogin() {
	process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

	const config = vscode.workspace.getConfiguration('komrad-filter-editor');
	const host = config.get('targetUrl');
	const loginPath = '/api/v1/login';
	// eslint-disable-next-line @typescript-eslint/naming-convention
	const loginCreds = { Login: config.get('login'), Password: config.get('password') };

	const loginUrl = host + loginPath;

	let response = await fetch(loginUrl, {
		method: "POST",
		body: JSON.stringify(loginCreds),
		headers: {
			// eslint-disable-next-line @typescript-eslint/naming-convention
			"Content-Type": "application/json",
		}
	});

	if (response.ok) {
		console.log('Login successful');
	}
	else {
		console.log('Login failed');
		return null;
	}

	var cookie = "something";

	var param = response.headers.get('set-cookie');
	if (param !== null) {
		var cookies = param.match(/SessionToken=(.*?);/);
		if (cookies !== null) {
			cookie = cookies[1];
		}
	}

	return cookie;
}

async function komradLogout(cookie: string) {

	const config = vscode.workspace.getConfiguration('komrad-filter-editor');
	const host = config.get('targetUrl');
	const logoutPath = '/api/v1/logout';

	const logoutUrl = host + logoutPath;
	let response = await fetch(logoutUrl, {
		method: "POST",
		headers: {
			// eslint-disable-next-line @typescript-eslint/naming-convention
			"Content-Type": "application/json",
			cookie: "SessionToken=" + cookie + ";"
		}
	});

	if (response.ok) {
		console.log('Logout successful');
	}
	else {
		console.log('Logout failed');
	}
}

async function getFiltersList(cookie: string) {
	process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

	const config = vscode.workspace.getConfiguration('komrad-filter-editor');
	const host = config.get('targetUrl');
	const getFiltersPath = '/api/queries?$sort=CreatedAt&$order=desc&$page=1&$limit=1000&$format=list';

	// let cookie = await komradLogin();

	const getFiltersUrl = host + getFiltersPath;

	let response = await fetch(getFiltersUrl, {
		method: "GET",
		headers: {
			// eslint-disable-next-line @typescript-eslint/naming-convention
			"Content-Type": "application/json",
			// eslint-disable-next-line @typescript-eslint/naming-convention
			"Accept": "application/json",
			"cookie": "SessionToken=" + cookie + ";"
		}
	});

	let filters = [];

	if (response.ok) {
		// console.log('Seems request works');
		let data = await response.json();
		// console.log(data.data.Items);
		for (let ind in data['data']['Items']) {
			let filter = new Filter(
				data['data']['Items'][ind]['Name'],
				data['data']['Items'][ind]['ID'],
				data['data']['Items'][ind]['Source'],
				data['data']['Items'][ind]['CreatedAt'],
				data['data']['Items'][ind]['Description']
			);
			filters.push(filter);
		}
		vscode.window.showInformationMessage('Фильтры успешно получены');
	}
	else {
		// console.log('Seems something went wrong');
		vscode.window.showErrorMessage('Не удалось получить фильтры');
		filters = [
			new Filter('name_1', 'id_1', 'source_1', '111', 'desc_1'),
			new Filter('name_2', 'id_2', 'source_2', '222', 'desc_2')
		];
	}

	return filters;
}

async function updateFilter(cookie: string, filter: Filter) {
	process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

	const config = vscode.workspace.getConfiguration('komrad-filter-editor');
	const host = config.get('targetUrl');
	const patchFiltersPath = '/api/queries/';

	const patchFiltersUrl = host + patchFiltersPath + filter.id;

	const editor = vscode.window.activeTextEditor;
	if (editor) {
		const data = {
			// eslint-disable-next-line @typescript-eslint/naming-convention
			CreatedAt: filter.createdAt,
			// eslint-disable-next-line @typescript-eslint/naming-convention
			Description: `${filter.description}`,
			// eslint-disable-next-line @typescript-eslint/naming-convention
			ID: `${filter.id}`,
			// eslint-disable-next-line @typescript-eslint/naming-convention
			Name: `${filter.label}`,
			// eslint-disable-next-line @typescript-eslint/naming-convention
			RuleTree: null,
			// eslint-disable-next-line @typescript-eslint/naming-convention
			Source: Buffer.from(editor.document.getText(), 'utf-8').toString('base64'),
			// eslint-disable-next-line @typescript-eslint/naming-convention
			SourceType: "lua"
		};

		let response = await fetch(patchFiltersUrl, {
			method: "PATCH",
			headers: {
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Type': 'application/json',
				// eslint-disable-next-line @typescript-eslint/naming-convention
				Accept: 'application/json',
				"cookie": "SessionToken=" + cookie + ";"
			},
			body: JSON.stringify(data)
		});

		if (response.ok) {
			vscode.window.showInformationMessage('Фльтр успешно обновлен. Обновите список фильтров.');
		}
		else {
			console.log(await response.statusText);
			vscode.window.showErrorMessage('Не удалось обновить фильтр! Повторите попытку.');
		}
	}
}

class Filter extends vscode.TreeItem {

	id: string;
	source: string;
	createdAt: string;
	description: string;

	constructor(label: string, id: string, source: string, createdAt: string, description: string) {
		super(label);
		this.id = id;
		this.source = source;
		this.createdAt = createdAt;
		this.description = description;
	}

}

export class FilterProvider implements vscode.TreeDataProvider<Filter> {

	filters: Filter[];

	constructor(filters: Filter[]) {
		this.filters = filters;
	}

	private _onDidChangeTreeData: vscode.EventEmitter<void | Filter | Filter[] | null | undefined> = new vscode.EventEmitter<void | Filter | Filter[] | null | undefined>();
	readonly onDidChangeTreeData: vscode.Event<void | Filter | Filter[] | null | undefined> = this._onDidChangeTreeData.event;

	async refresh(cookie: string) {
		this.filters = await getFiltersList(cookie);
		this._onDidChangeTreeData.fire();
	}

	getTreeItem(element: Filter): vscode.TreeItem | Thenable<vscode.TreeItem> {
		return element;
	}

	getChildren(element?: Filter | undefined): vscode.ProviderResult<Filter[]> {
		if (element === undefined) {
			return this.filters;
		}
		else {
			return null;
		}
	}

	getParent?(element: Filter): vscode.ProviderResult<Filter> {
		throw new Error('Method not implemented.');
	}

	resolveTreeItem?(item: vscode.TreeItem, element: Filter, token: vscode.CancellationToken): vscode.ProviderResult<vscode.TreeItem> {
		throw new Error('Method not implemented.');
	}

}

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterProvider = exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const node_fetch_1 = require("node-fetch");
async function activate(context) {
    console.log('Congratulations, your extension "komrad-filter-editor" is now active!');
    const config = vscode.workspace.getConfiguration('komrad-filter-editor');
    let cookie = config.get('token');
    if (!cookie || cookie === undefined || cookie === null) {
        cookie = await komradLogin();
    }
    if (cookie !== null) {
        const filterProvider = new FilterProvider(await getFiltersList(String(cookie)));
        let filterTree = vscode.window.registerTreeDataProvider('filters-list', filterProvider);
        let filtersList = vscode.commands.registerCommand('filters-list.refresh', () => {
            filterProvider.refresh(String(cookie));
        });
        let filterEdit = vscode.commands.registerCommand('filters-list.edit', (node) => {
            let editor = vscode.window.activeTextEditor;
            let sourceCode = Buffer.from(node.source, 'base64').toString('utf-8');
            if (editor) {
                let range = new vscode.Range(0, 0, editor.document.lineCount, 0);
                editor.edit(editBuilder => {
                    editBuilder.replace(range, sourceCode);
                });
                vscode.window.showInformationMessage(`Фльтр ${node.label} успшно получен`);
            }
            else {
                vscode.window.showErrorMessage('Сначала откроте новое пустое окно (Ctrl+N)');
            }
        });
        let filterUpload = vscode.commands.registerCommand('filters-list.upload', (node) => {
            updateFilter(String(cookie), node);
        });
        context.subscriptions.push(filterTree, filtersList, filterEdit, filterUpload);
    }
}
exports.activate = activate;
async function deactivate() { }
exports.deactivate = deactivate;
async function komradLogin() {
    process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
    const config = vscode.workspace.getConfiguration('komrad-filter-editor');
    const host = config.get('targetUrl');
    const loginPath = '/api/v1/login';
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const loginCreds = { Login: config.get('login'), Password: config.get('password') };
    const loginUrl = host + loginPath;
    let response = await (0, node_fetch_1.default)(loginUrl, {
        method: "POST",
        body: JSON.stringify(loginCreds),
        headers: {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            "Content-Type": "application/json",
        }
    });
    if (response.ok) {
        console.log('Login successful');
    }
    else {
        console.log('Login failed');
        return null;
    }
    var cookie = "something";
    var param = response.headers.get('set-cookie');
    if (param !== null) {
        var cookies = param.match(/SessionToken=(.*?);/);
        if (cookies !== null) {
            cookie = cookies[1];
        }
    }
    return cookie;
}
async function komradLogout(cookie) {
    const config = vscode.workspace.getConfiguration('komrad-filter-editor');
    const host = config.get('targetUrl');
    const logoutPath = '/api/v1/logout';
    const logoutUrl = host + logoutPath;
    let response = await (0, node_fetch_1.default)(logoutUrl, {
        method: "POST",
        headers: {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            "Content-Type": "application/json",
            cookie: "SessionToken=" + cookie + ";"
        }
    });
    if (response.ok) {
        console.log('Logout successful');
    }
    else {
        console.log('Logout failed');
    }
}
async function getFiltersList(cookie) {
    process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
    const config = vscode.workspace.getConfiguration('komrad-filter-editor');
    const host = config.get('targetUrl');
    const getFiltersPath = '/api/queries?$sort=CreatedAt&$order=desc&$page=1&$limit=1000&$format=list';
    // let cookie = await komradLogin();
    const getFiltersUrl = host + getFiltersPath;
    let response = await (0, node_fetch_1.default)(getFiltersUrl, {
        method: "GET",
        headers: {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            "Content-Type": "application/json",
            // eslint-disable-next-line @typescript-eslint/naming-convention
            "Accept": "application/json",
            "cookie": "SessionToken=" + cookie + ";"
        }
    });
    let filters = [];
    if (response.ok) {
        // console.log('Seems request works');
        let data = await response.json();
        // console.log(data.data.Items);
        for (let ind in data['data']['Items']) {
            let filter = new Filter(data['data']['Items'][ind]['Name'], data['data']['Items'][ind]['ID'], data['data']['Items'][ind]['Source'], data['data']['Items'][ind]['CreatedAt'], data['data']['Items'][ind]['Description']);
            filters.push(filter);
        }
        vscode.window.showInformationMessage('Фильтры успешно получены');
    }
    else {
        // console.log('Seems something went wrong');
        vscode.window.showErrorMessage('Не удалось получить фильтры');
        filters = [
            new Filter('name_1', 'id_1', 'source_1', '111', 'desc_1'),
            new Filter('name_2', 'id_2', 'source_2', '222', 'desc_2')
        ];
    }
    return filters;
}
async function updateFilter(cookie, filter) {
    process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
    const config = vscode.workspace.getConfiguration('komrad-filter-editor');
    const host = config.get('targetUrl');
    const patchFiltersPath = '/api/queries/';
    const patchFiltersUrl = host + patchFiltersPath + filter.id;
    const editor = vscode.window.activeTextEditor;
    if (editor) {
        const data = {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            CreatedAt: filter.createdAt,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            Description: `${filter.description}`,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            ID: `${filter.id}`,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            Name: `${filter.label}`,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            RuleTree: null,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            Source: Buffer.from(editor.document.getText(), 'utf-8').toString('base64'),
            // eslint-disable-next-line @typescript-eslint/naming-convention
            SourceType: "lua"
        };
        let response = await (0, node_fetch_1.default)(patchFiltersUrl, {
            method: "PATCH",
            headers: {
                // eslint-disable-next-line @typescript-eslint/naming-convention
                'Content-Type': 'application/json',
                // eslint-disable-next-line @typescript-eslint/naming-convention
                Accept: 'application/json',
                "cookie": "SessionToken=" + cookie + ";"
            },
            body: JSON.stringify(data)
        });
        if (response.ok) {
            vscode.window.showInformationMessage('Фльтр успешно обновлен. Обновите список фильтров.');
        }
        else {
            console.log(await response.statusText);
            vscode.window.showErrorMessage('Не удалось обновить фильтр! Повторите попытку.');
        }
    }
}
class Filter extends vscode.TreeItem {
    constructor(label, id, source, createdAt, description) {
        super(label);
        this.id = id;
        this.source = source;
        this.createdAt = createdAt;
        this.description = description;
    }
}
class FilterProvider {
    constructor(filters) {
        this._onDidChangeTreeData = new vscode.EventEmitter();
        this.onDidChangeTreeData = this._onDidChangeTreeData.event;
        this.filters = filters;
    }
    async refresh(cookie) {
        this.filters = await getFiltersList(cookie);
        this._onDidChangeTreeData.fire();
    }
    getTreeItem(element) {
        return element;
    }
    getChildren(element) {
        if (element === undefined) {
            return this.filters;
        }
        else {
            return null;
        }
    }
    getParent(element) {
        throw new Error('Method not implemented.');
    }
    resolveTreeItem(item, element, token) {
        throw new Error('Method not implemented.');
    }
}
exports.FilterProvider = FilterProvider;
//# sourceMappingURL=extension.js.map